File: cpp/src/v0/book.cpp
#include "book.hpp"

#include <algorithm>

namespace ytta::v0 {

void Book::add(const Order& order) {
  BookLevelOrder level{order.cl_ord_id, order.qty};
  if (order.side == Side::Buy) {
    bids_[order.price].push_back(level);
  } else {
    asks_[order.price].push_back(level);
  }
  side_by_id_[order.cl_ord_id] = order.side;
}

std::optional<std::int64_t> Book::cancel(std::uint64_t cl_ord_id) {
  const auto side_it = side_by_id_.find(cl_ord_id);
  if (side_it == side_by_id_.end()) {
    return std::nullopt;
  }
  const Side side = side_it->second;
  auto erase_from = [&](auto& levels) -> std::optional<std::int64_t> {
    for (auto lit = levels.begin(); lit != levels.end(); ++lit) {
      auto& q = lit->second;
      for (auto oit = q.begin(); oit != q.end(); ++oit) {
        if (oit->cl_ord_id == cl_ord_id) {
          const std::int64_t qty = oit->qty;
          q.erase(oit);
          if (q.empty()) {
            levels.erase(lit);
          }
          side_by_id_.erase(cl_ord_id);
          return qty;
        }
      }
    }
    return std::nullopt;
  };
  if (side == Side::Buy) {
    return erase_from(bids_);
  }
  return erase_from(asks_);
}

std::vector<Book::MatchSlice> Book::match(Side aggressor_side,
                                          std::int64_t price,
                                          std::int64_t qty) {
  std::vector<MatchSlice> slices;
  std::int64_t remaining = qty;

  auto match_levels = [&](auto& levels, auto crosses) {
    while (remaining > 0 && !levels.empty()) {
      auto lit = levels.begin();
      if (!crosses(lit->first, price)) {
        break;
      }
      auto& q = lit->second;
      while (remaining > 0 && !q.empty()) {
        auto& head = q.front();
        const std::int64_t fill_qty = std::min(remaining, head.qty);
        slices.push_back(MatchSlice{head.cl_ord_id, lit->first, fill_qty});
        head.qty -= fill_qty;
        remaining -= fill_qty;
        if (head.qty == 0) {
          side_by_id_.erase(head.cl_ord_id);
          q.pop_front();
        }
      }
      if (q.empty()) {
        levels.erase(lit);
      }
    }
  };

  if (aggressor_side == Side::Buy) {
    match_levels(asks_, [](std::int64_t ask_px, std::int64_t buy_px) {
      return ask_px <= buy_px;
    });
  } else {
    match_levels(bids_, [](std::int64_t bid_px, std::int64_t sell_px) {
      return bid_px >= sell_px;
    });
  }
  return slices;
}

MdEvent Book::snapshot(std::int64_t ts_ns) const {
  MdEvent md;
  md.ts_ns = ts_ns;
  md.instrument = "INSTR1";
  if (!bids_.empty()) {
    md.best_bid = bids_.begin()->first;
    std::int64_t qty = 0;
    for (const auto& o : bids_.begin()->second) {
      qty += o.qty;
    }
    md.best_bid_qty = qty;
  }
  if (!asks_.empty()) {
    md.best_ask = asks_.begin()->first;
    std::int64_t qty = 0;
    for (const auto& o : asks_.begin()->second) {
      qty += o.qty;
    }
    md.best_ask_qty = qty;
  }
  return md;
}

}  // namespace ytta::v0

File: rust/src/v0/book.rs
use std::collections::{BTreeMap, HashMap, VecDeque};

use super::types::{MdEvent, Order, Side};

#[derive(Clone)]
struct BookLevelOrder {
    cl_ord_id: u64,
    qty: i64,
}

pub struct MatchSlice {
    pub maker_id: u64,
    pub price: i64,
    pub qty: i64,
}

pub struct Book {
    bids: BTreeMap<i64, VecDeque<BookLevelOrder>>,
    asks: BTreeMap<i64, VecDeque<BookLevelOrder>>,
    side_by_id: HashMap<u64, Side>,
}

impl Book {
    pub fn new() -> Self {
        Self {
            bids: BTreeMap::new(),
            asks: BTreeMap::new(),
            side_by_id: HashMap::new(),
        }
    }

    pub fn add(&mut self, order: Order) {
        let level = BookLevelOrder {
            cl_ord_id: order.cl_ord_id,
            qty: order.qty,
        };
        match order.side {
            Side::Buy => self.bids.entry(order.price).or_default().push_back(level),
            Side::Sell => self.asks.entry(order.price).or_default().push_back(level),
        }
        self.side_by_id.insert(order.cl_ord_id, order.side);
    }

    pub fn cancel(&mut self, cl_ord_id: u64) -> Option<i64> {
        let side = *self.side_by_id.get(&cl_ord_id)?;
        let levels = match side {
            Side::Buy => &mut self.bids,
            Side::Sell => &mut self.asks,
        };
        let mut found_px = None;
        let mut found_qty = None;
        for (px, q) in levels.iter_mut() {
            if let Some(pos) = q.iter().position(|o| o.cl_ord_id == cl_ord_id) {
                let removed = q.remove(pos).unwrap();
                found_qty = Some(removed.qty);
                found_px = Some(*px);
                break;
            }
        }
        let px = found_px?;
        let qty = found_qty?;
        if levels.get(&px).map(|q| q.is_empty()).unwrap_or(true) {
            levels.remove(&px);
        }
        self.side_by_id.remove(&cl_ord_id);
        Some(qty)
    }

    pub fn match_order(&mut self, aggressor: Side, price: i64, qty: i64) -> Vec<MatchSlice> {
        let mut slices = Vec::new();
        let mut remaining = qty;

        match aggressor {
            Side::Buy => {
                while remaining > 0 {
                    let best = match self.asks.keys().next().copied() {
                        Some(px) if px <= price => px,
                        _ => break,
                    };
                    Self::match_level(
                        &mut self.asks,
                        &mut self.side_by_id,
                        best,
                        &mut remaining,
                        &mut slices,
                    );
                }
            }
            Side::Sell => {
                while remaining > 0 {
                    let best = match self.bids.keys().next_back().copied() {
                        Some(px) if px >= price => px,
                        _ => break,
                    };
                    Self::match_level(
                        &mut self.bids,
                        &mut self.side_by_id,
                        best,
                        &mut remaining,
                        &mut slices,
                    );
                }
            }
        }
        slices
    }

    fn match_level(
        levels: &mut BTreeMap<i64, VecDeque<BookLevelOrder>>,
        side_by_id: &mut HashMap<u64, Side>,
        px: i64,
        remaining: &mut i64,
        slices: &mut Vec<MatchSlice>,
    ) {
        let q = levels.get_mut(&px).unwrap();
        while *remaining > 0 && !q.is_empty() {
            let head = q.front_mut().unwrap();
            let fill_qty = (*remaining).min(head.qty);
            slices.push(MatchSlice {
                maker_id: head.cl_ord_id,
                price: px,
                qty: fill_qty,
            });
            head.qty -= fill_qty;
            *remaining -= fill_qty;
            if head.qty == 0 {
                let done = q.pop_front().unwrap();
                side_by_id.remove(&done.cl_ord_id);
            }
        }
        if q.is_empty() {
            levels.remove(&px);
        }
    }

    pub fn snapshot(&self, ts_ns: i64) -> MdEvent {
        let mut md = MdEvent {
            ts_ns,
            instrument: "INSTR1".to_string(),
            best_bid: 0,
            best_ask: 0,
            best_bid_qty: 0,
            best_ask_qty: 0,
        };
        if let Some((&px, q)) = self.bids.iter().next_back() {
            md.best_bid = px;
            md.best_bid_qty = q.iter().map(|o| o.qty).sum();
        }
        if let Some((&px, q)) = self.asks.iter().next() {
            md.best_ask = px;
            md.best_ask_qty = q.iter().map(|o| o.qty).sum();
        }
        md
    }
}

impl Default for Book {
    fn default() -> Self {
        Self::new()
    }
}
